hack/build_channel_catalog.sh

```diff
@@ -58,52 +62,43 @@ else
+log "Cloning $BUNDLE_VERSIONS_REPO into $saas_operator_dir_base"
+git clone --branch "$BRANCH" "$bundle_versions_repo_url" "$saas_operator_dir_base"
 
-prev_operator_version=""
-removed_versions=""
+# if the line contains SKIP it will not be included
 if [[ -s "$bundle_versions_file" ]]; then
     log "$bundle_versions_file exists. We'll use to determine current version"
-    if [[ "$REMOVE_UNDEPLOYED" == "true" ]]; then
-        log "Checking if we have to remove any versions more recent than deployed hash"
-        # this will need to be modified if at any point we move to a canary like deployment
-        # where not all environments are deployed with the same operator version
-        deployed_hash=$(
-            curl -s -H "Authorization: Basic $(echo -n $APP_INTERFACE_USERNAME:$APP_INTERFACE_PASSWORD | base64)" \
-                -g "https://$APP_INTERFACE_BASE_URL/graphql?query={saas_files:saas_files_v1{name,resourceTemplates{name,targets{namespace{environment{name,labels}},ref}}}}" | \
-                jq -r '.data.saas_files[] | select(.name=="saas-'$OPERATOR_NAME'") | .resourceTemplates[].targets[] | select(.namespace.environment.labels | contains("\"type\":\"production\"")) | .ref' | \
-                uniq
-        )
-
-        log "Current deployed hash is $deployed_hash"
-
-        new_bundle_versions_file=$(mktemp -p "$temp_dir" new_bundle_versions_file.XXXX)
-        delete=false
-        # Sort based on commit number
-        for version in $(sort -t . -k 3 -g "$bundle_versions_file"); do
-            if [[ "$delete" == false ]]; then
-                echo "$version" >> "$new_bundle_versions_file"
-
-                short_hash=$(echo "$version" | cut -d- -f2)
-
-                if [[ "$deployed_hash" == "${short_hash}"* ]]; then
-                    log "found deployed hash in the bundle versions repository"
-                    delete=true
-                fi
-            else
-                log "Adding $version to removed versions"
-                removed_versions="$version $removed_versions"
-            fi
-        done
-        [[ -n "$removed_versions" ]] && log "The following versions will be removed: $removed_versions"
-        cp "$new_bundle_versions_file" "$bundle_versions_file"
+
+    prev_operator_version=$(tail -n 1 "$bundle_versions_file" | awk '{print $1}')
+    prev_good_operator_version=""
+    skip_versions=()
+
+    # we traverse the bundle versions file backwards
+    # we cannot use pipes here or we would lose the inner variables changes
+    while read -r line; do
+        if [[ $(echo "$line" | grep SKIP) == "" ]]; then
+            prev_good_operator_version="$line"
+            break
+        else
+            version=$(echo "$line" | awk '{print $1}')
+            skip_versions+=("$version")
+        fi
+    done < <(sort -r -t . -k 3 -g "$bundle_versions_file")
+
+    if [[ -z "$prev_good_operator_version" ]]; then
```

Comment:
Oh, obviates the corner case I mentioned below. Not sure it's legit to bail here, though. If the versions file contains only

```
v1.0.0 SKIP
```
...don't we want to use `skipRange` but no `replaces`?

We should probably ask the OLM folks on this.

Reply:
you're indeed correct, but If you want to skip your only version... you may want to start from scratch.

The idea was to continue using `replaces`, I thought there was an agreement on that since it's one of the foundations of this PR.

Reply:
If the versions file only contains a SKIP entry then we will be creating a new catalog from scratch.  There shouldn't be a replaces used at all in that case.

Reply:
Summary of that discussion:
Specifically in appsre it doesn't matter because the process of purging bad versions is manual.
but
In general, setting `replaces` to the latest skipped version -- the only time `replaces` and `skipRange` overlap -- would provide a path from the bad version if it's already installed somewhere, but also ensure that that version isn't available for fresh installation.
so
IMO we should do the latter anyway, since it's harmless for appsre but more generally correct.

Reply:
Apologies in advance, but I don't understand what you mean here, @2uasimojo 😅  I got lost in _"In general, setting replaces to the latest skipped version"_ as it is something that we avoid with the `SKIP` logic

Reply:
> Apologies in advance, but I don't understand what you mean here, @2uasimojo  I got lost in _"In general, setting replaces to the latest skipped version"_ as it is something that we avoid with the `SKIP` logic

Sorry, my "in general" was _too_ general.

I meant that, in the corner case where we need to skip all existing versions, we can set `replaces` to the latest skipped version. That's the only time `replaces` has a value covered by `skipRange`. This would be the correct thing to do for this corner case in The World (i.e. outside of appsre).

Specifically in appsre it doesn't matter, since we're purging the bad versions manually.

Reply:
In the case we need to skip all existing versions, we just end up with a new catalog.  There's no need to set skip for our use case, and in fact setting skip/replaces for a catalog w/o the versions being skipped/replaced doesn't make sense and might cause problems.  I don't think this is something we need to worry about.  Current process will handle this case with a new catalog.  We only allow good version in a catalog.

Reply:
Clarified. Let's keep `replaces`, and let's fix the edge case by adding in the replaces the last SKIP version in that very specific case to avoid manual intervention if possible.

Reply:
fixed in 989a514

Reply:
✓